src/System.Private.CoreLib/src/System/Threading/Timer.cs

```diff
@@ -210,138 +200,192 @@ internal static void AppDomainTimerCallback(int id)
-                            // to the ThreadPool.
-                            //
+                            // If this is the first timer, we'll fire it on this thread (after processing
+                            // all others). Otherwise, queue it to the ThreadPool.
                             if (timerToFireOnThisThread == null)
+                            {
                                 timerToFireOnThisThread = timer;
+                            }
                             else
-                                QueueTimerCompletion(timer);
-
-                            timer = nextTimer;
+                            {
+                                ThreadPool.UnsafeQueueCustomWorkItem(timer, forceGlobal: true);
+                            }
                         }
                         else
                         {
-                            //
-                            // This timer hasn't fired yet.  Just update the next time the native timer fires.
-                            //
-                            uint remaining = timer.m_dueTime - elapsed;
+                            // This timer isn't ready to fire.  Update the next time the native timer fires if necessary,
+                            // and move this timer to the short list if its remaining time is now at or under the threshold.
+
                             if (remaining < nextAppDomainTimerDuration)
                             {
                                 haveTimerToSchedule = true;
-                                nextAppDomainTimerDuration = remaining;
+                                nextAppDomainTimerDuration = (uint)remaining;
+                            }
+
+                            if (!timer.m_short && remaining <= ShortTimersThresholdMilliseconds)
+                            {
+                                MoveTimerToCorrectList(timer, shortList: true);
                             }
-                            timer = timer.m_next;
                         }
+
+                        timer = next;
                     }
 
-                    if (haveTimerToSchedule)
-                        EnsureAppDomainTimerFiresBy(nextAppDomainTimerDuration);
+                    // If we just finished procesing the short list, we can stop and avoid the long list if
+                    // we're <= the current threshold by which timers were partitioned into the short list
+                    // and the short list is not empty (if it's empty, we need to evaluate the long list to
+                    // know when to next schedule the appdomain timer). Otherwise, we move on to processing
+                    // the long list.
+                    if (listNum == 0)
+                    {
+                        bool processLongList = startAbsoluteThreshold - nowTicks < 0 || m_shortTimers == null;
+                        if (!processLongList)
```

Comment:
This design is simple, but in the following scenario.
* You have MANY timeout timers (say 10  second durations).
* You have threads that keeps creating a short delays (e.g. Delay(1)).  

Under such conditions, every time the Fire method happens (which is on every tick, since the delay is short), you end up having to look at the long list (to determine the next fire time).  

You could avoid this by either 
1. Tracking the minimum of the long list.   OR
2. Simply setting  nextAppDomainTimerDuration to what is left between now and startAbsoluteThreshold.  

By delaying you give time for either more short timers to arrive, or for other timeout timers to die.   But what is nice is that you can say that you will never look at the long list more frequently than once every 333 msec, which is a very nice thing to be able to say.

Reply:
(2) seems simple and reasonable.

Reply:
I agree that one is very easy and will get the job done.

Reply:
I added a commit with that change.